proposals/handler.go

```diff
@@ -0,0 +1,380 @@
+		return nil
+	}
+	logger.With().Info("new ballot", b.Fields()...)
+
+	if err := h.checkBallotSyntacticValidity(ctx, b); err != nil {
+		logger.With().Error("ballot syntactically invalid", log.Err(err))
+		return fmt.Errorf("syntactic-check ballot: %w", err)
+	}
+
+	reportBallotMetrics(b)
+	return nil
+}
+
+func (h *Handler) checkBallotSyntacticValidity(ctx context.Context, b *types.Ballot) error {
+	h.logger.WithContext(ctx).With().Debug("checking proposal syntactic validity")
+
+	if err := h.checkBallotDataIntegrity(b); err != nil {
+		h.logger.WithContext(ctx).With().Error("ballot integrity check failed", log.Err(err))
+		return err
+	}
+
+	if err := h.checkBallotDataAvailability(ctx, b); err != nil {
+		h.logger.WithContext(ctx).With().Error("ballot data availability check failed", log.Err(err))
+		return err
+	}
+
+	if eligible, err := h.validator.CheckEligibility(ctx, b); err != nil || !eligible {
+		h.logger.WithContext(ctx).With().Error("ballot eligibility check failed", log.Err(err))
+		return errNotEligible
+	}
+
+	h.logger.WithContext(ctx).With().Debug("ballot is syntactically valid")
+	return nil
+}
+
+func (h *Handler) checkBallotDataIntegrity(b *types.Ballot) error {
+	if b.AtxID == *types.EmptyATXID || b.AtxID == h.cfg.GoldenATXID {
+		return errInvalidATXID
+	}
+
+	if b.BaseBallot == types.EmptyBallotID {
+		return errMissingBaseBallot
+	}
+
+	if b.RefBallot == types.EmptyBallotID {
+		// this is the smesher's first Ballot in this epoch, should contain EpochData
+		if b.EpochData == nil {
+			return errMissingEpochData
+		}
+		if len(b.EpochData.Beacon) == 0 {
+			return errMissingBeacon
+		}
+		if len(b.EpochData.ActiveSet) == 0 {
+			return errEmptyActiveSet
+		}
+		// check for duplicate ATXIDs in active set
+		set := make(map[types.ATXID]struct{}, len(b.EpochData.ActiveSet))
+		for _, atx := range b.EpochData.ActiveSet {
+			if _, exist := set[atx]; exist {
+				return errDuplicateATX
+			}
+			set[atx] = struct{}{}
+		}
+	} else {
+		if b.EpochData != nil {
+			return errUnexpectedEpochData
+		}
+	}
+
+	if err := h.checkExceptions(b, h.cfg.MaxExceptions); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (h *Handler) checkExceptions(ballot *types.Ballot, max int) error {
+	exceptions := map[types.BlockID]struct{}{}
+	for _, diff := range [][]types.BlockID{ballot.ForDiff, ballot.NeutralDiff, ballot.AgainstDiff} {
+		for _, bid := range diff {
+			_, exist := exceptions[bid]
+			if exist {
+				return fmt.Errorf("%w: block %s is referenced multiple times in exceptions of ballot %s",
+					errConflictingExceptions, bid, ballot.ID())
+			}
+			exceptions[bid] = struct{}{}
+		}
+	}
+	if len(exceptions) > max {
+		return fmt.Errorf("%w: %d exceptions with max allowed %d in ballot %s",
+			errExceptionsOverflow, len(exceptions), max, ballot.ID())
+	}
+	return nil
+}
+
+func ballotBlockView(b *types.Ballot) []types.BlockID {
+	combined := make([]types.BlockID, 0, len(b.AgainstDiff)+len(b.ForDiff)+len(b.NeutralDiff))
```

Comment:
why base ballot was removed from this list and now fetched separately? can we join everything then into a single list and fetch in a single call

Reply:
base ballot is now fetched together with ref ballot if any. the separation is done because we have different API for fetching Ballot (`GetBallots()`) and Block (`GetBlocks()`).